feat(transformer): support exponentiation operator plugin

Dunqing opened the pull request:
The implementation copy from the original implementation which removed in https://github.com/oxc-project/oxc/pull/2865.

Dunqing commented:
* **#4876** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4876?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_transformer/src/es2016/exponentiation_operator.rs:
```diff
@@ -0,0 +1,322 @@
+        if let Expression::AssignmentExpression(assign_expr) = expr {
+            if assign_expr.operator == AssignmentOperator::Exponential {
+                let mut nodes = ctx.ast.vec();
+                let Some(Exploded { reference, uid }) =
+                    self.explode(&mut assign_expr.left, &mut nodes, ctx)
+                else {
+                    return;
+                };
```
Where left hand side is a member expression, I think we can avoid cloning it. We could instead get an owned copy of `left` here with `ast.move_expression(&assign_expr.left)`. Then insert that in new position, with no cloning.

```js
// In
obj.prop **= exp;

// Out
// NB: only 1 x `obj` - no cloning needed
var _obj;
(_obj = obj), (_obj.prop = Math.pow(_obj.prop, exp));
```

In above example, there's only 1 `obj` in the output, so no need for cloning. I think this also holds for all `MemberExpression` cases.

In the simpler case where left hand side is just an identifier, we *do* need to clone it:

```js
// In
value **= exp;

// Out
// NB: 2 x `value`s - cloning needed
var _value;
(_value = value), (value = Math.pow(_value, exp))
```

but cloning an identifier is much simpler than cloning an expression.

overlookmotel commented on crates/oxc_transformer/src/es2016/exponentiation_operator.rs:
```diff
@@ -0,0 +1,322 @@
+        // left ** right
+        if let Expression::BinaryExpression(binary_expr) = expr {
+            if binary_expr.operator == BinaryOperator::Exponential {
+                let left = ctx.ast.move_expression(&mut binary_expr.left);
+                let right = ctx.ast.move_expression(&mut binary_expr.right);
```
Thank you for not using `ast.copy` here! ❤️

Dunqing replied:
Haha, If I still use `ast.copy`, it would cause tons of semantic errors after #4790

overlookmotel commented:
As discussed earlier, I wasn't planning to review these new transform PRs at this stage. I proposed that we merge them and then I do a second pass to review/tidy them later on. I only looked into this one because it informed #4880.

Dunqing commented:
> As discussed earlier, I wasn't planning to review these new transform PRs at this stage. I proposed that we merge them and then I do a second pass to review/tidy them later on. I only looked into this one because it informed #4880.

Let's merge all this stack of PRs first. We can make many improvements later

Dunqing commented:
> As discussed earlier, I wasn't planning to review these new transform PRs at this stage. I proposed that we merge them and then I do a second pass to review/tidy them later on. I only looked into this one because it informed #4880.
 
Thank you for your review, Let's merge all this stack of PRs first. We can make many improvements later

Dunqing committed:
feat(transformer): support exponentiation operator plugin (#4876)

The implementation copy from the original implementation which removed in https://github.com/oxc-project/oxc/pull/2865.
